include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp

```diff
@@ -57,12 +57,30 @@ namespace alpaka
                     false>; // If the threads should yield.
 
             public:
-                explicit QueueGenericThreadsNonBlockingImpl(TDev dev) : m_dev(std::move(dev)), m_workerThread(1u)
+                explicit QueueGenericThreadsNonBlockingImpl(TDev dev)
+                    : m_dev(std::move(dev))
+                    , m_workerThread(std::make_unique<ThreadPool>(1u))
                 {
                 }
                 QueueGenericThreadsNonBlockingImpl(QueueGenericThreadsNonBlockingImpl<TDev> const&) = delete;
+                QueueGenericThreadsNonBlockingImpl(QueueGenericThreadsNonBlockingImpl<TDev>&&) = delete;
                 auto operator=(QueueGenericThreadsNonBlockingImpl<TDev> const&)
                     -> QueueGenericThreadsNonBlockingImpl<TDev>& = delete;
+                auto operator=(QueueGenericThreadsNonBlockingImpl&&)
+                    -> QueueGenericThreadsNonBlockingImpl<TDev>& = delete;
+                ~QueueGenericThreadsNonBlockingImpl() override
+                {
+                    m_dev.registerCleanup(
+                        [pool = std::weak_ptr<ThreadPool>(m_workerThread)]() noexcept
+                        {
+                            auto s = pool.lock();
+                            if(s)
+                            {
+                                s = s->takeDetachHandle();
+                            }
+                        });
+                    m_workerThread->detach(std::move(m_workerThread));
```

Comment:
Is this legal ?
`detach()` is invoked on `m_workerThread` _after_ is has been moved from.

Reply:
Yes.

All arguments for the function are obtained first (i.e. `this` for `detach()` and the rvalue smart pointer) in no specific order, but before the function `detach` is invoked. `std::move` only casts the reference to the object into an rvalue reference (basically it only affects overload resolution), it does not modify the object. The actual change to `m_workerThread`, which puts it into a moved-from state, happens inside `detach`, in for of the move-assignment.

Reply:
Also note, that `detach()` is not invoked on`m_workerThread`, which is a `shared_ptr<>` (in the latest version), but on `*m_workerThread`, which is a `ConcurrentExecPool`.

Reply:
Thanks for the explanation, I tend to agree that the actual "moving from" happens inside the function body, after the function call.

Just on one point:

> the `operator->` constitutes a sequence point,

I cannot find any reference to this being the case, could you point to it ?

Reply:
Hmm, maybe there is actually no sequence defined.

[Here](https://en.cppreference.com/w/cpp/language/eval_order) under "Rules", rules 2 or 14 could apply somehow, but 2 sequences only the value computations, so the side-effect of a move-assignment in the function-call expression could still go anywhere. I do not know, if, in 14, the "expression that names the function" includes the object and member access, or only anything that is where the function name goes, like some expression evaluation to a function pointer.

Reply:
**disclaimer**: *this is a contrived example and has nothing to do with the code in the PR*.

But I was curious, so I've tried to come up with a simpler example where the order of evaluation would actually have observable consequences: https://godbolt.org/z/j6c1GK8Tj .

I _think_ the order of evaluation in `p->print(move_from_and_return(p))` is the same as in `m_workerThread->detach(std::move(m_workerThread))`.

In the contrived example, the argument to the function _does_ "move from" the `shared_ptr`  _before_ the method is called, resulting in the call appling to a `nullptr`, and leading to a crash.

(In the PR, I agree that `std::move` does not modify the `shared_ptr` itself, so the code is fine).

Reply:
I also played with some example, but only with different ways of assigning the arguments (e.g. by-value instead of by-rvalue-ref), not with calling a function that has side-effects (`std::move` is only a cast). Intuitively, I think that the evaluation of the call to `move_from_and_return()` might even be guaranteed to happen before the outer parts of the expression, not just be called first depending on the compiler's whim.

I think we clarified, that `->` does not imply any sequence of evaluation here. I.e., if `detach` would take a value, than the compiler may do the move-assignment of the parameter or the dereference at `->` first as it chooses ... Or it maybe, since C++17, it is compelled to evaluate the part identifying the function first before assigning its argument, not sure.

Reply:
> Intuitively, I think that the evaluation of the call to `move_from_and_return()` might even be guaranteed to happen before the outer parts of the expression

Although, I do not see any of the rules implying this, so I guess the order in your example is probably undefined.